### PR TITLE
Torch model reader refactor

### PR DESCRIPTION
**Pull request by elad-c**
## Pull Request Description:
Refactor PyTorch model reader & builder to handle weights and tensors that are keyword inputs.

## Checklist before requesting a review:
- [x] I set the appropriate labels on the pull request.
- [x] I have added/updated the release note draft (if necessary).
- [x] I have updated the documentation to reflect my changes (if necessary).
- [x] All function and files are well documented. 
- [x] All function and classes have type hints.
- [x] There is a licenses in all file.
- [x] The function and variable names are informative. 
- [x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py [hunk 1]
```diff
@@ -66,31 +66,39 @@ def _build_input_tensors_list(node: BaseNode,
...
         _input_list = op_call_args.copy()
-        if tensor_input_indices is None:
-            tensor_input_indices = _node.tensor_input_indices
-        assert len(tensor_input_indices) == len(input_tensors), \
-            f'Mismatch between input tensors ({len(tensor_input_indices)}) and indices {len(input_tensors)}'
-        for i, t in zip(tensor_input_indices, input_tensors):
-            _input_list.insert(i, t)
+        if tensor_input_allocs is None:
+            tensor_input_allocs = _node.tensor_input_allocs
+        assert len(tensor_input_allocs) == len(input_tensors), \
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py:88]: why not logger?

File: model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py [hunk 2]
```diff
@@ -66,31 +66,39 @@ def _build_input_tensors_list(node: BaseNode,
...
         input_tensors: activation input tensors to node.
         op_call_args: framework node call args.
+        op_call_kwargs: framework node call kwargs.
+        tensor_input_allocs: List of input allocations to node.
 
     Returns:
         Combined list of input_tensors and op_call_args.
     """
-    if isinstance(_node, FunctionalNode) and _node.tensor_input_indices:
+    if isinstance(_node, FunctionalNode) and _node.tensor_input_allocs:
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py:84]: what happens if _node.tensor_input_allocs is None but the argument tensor_input_allocs is not None? is that case even possible?
if not, is `tensor_input_allocs == _node.tensor_input_allocs` always?

maybe at least explain what is the difference between them

File: model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py [hunk 3]
```diff
@@ -66,31 +66,39 @@ def _build_input_tensors_list(node: BaseNode,
...
-        if tensor_input_indices is None:
-            tensor_input_indices = _node.tensor_input_indices
-        assert len(tensor_input_indices) == len(input_tensors), \
-            f'Mismatch between input tensors ({len(tensor_input_indices)}) and indices {len(input_tensors)}'
-        for i, t in zip(tensor_input_indices, input_tensors):
-            _input_list.insert(i, t)
+        if tensor_input_allocs is None:
+            tensor_input_allocs = _node.tensor_input_allocs
+        assert len(tensor_input_allocs) == len(input_tensors), \
+            f'Mismatch between input tensors ({len(tensor_input_allocs)}) and indices {len(input_tensors)}'
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py:89]: add node name to comment (and add '.' at the end of the sentence)

File: model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py [hunk 4]
```diff
@@ -66,31 +66,39 @@ def _build_input_tensors_list(node: BaseNode,
...
+        for i, t in zip(tensor_input_allocs, input_tensors):
+            if isinstance(i, str):
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py:91]: add a comment explaining this if-else

File: model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py [hunk 5]
```diff
@@ -66,31 +66,39 @@ def _build_input_tensors_list(node: BaseNode,
     return input_tensors
 
 
-def _merge_inputs(_node: BaseNode, input_tensors: List, op_call_args: List,
-                  tensor_input_indices: List = None) -> List:
+def _merge_inputs(_node: BaseNode, input_tensors: List, op_call_args: List, op_call_kwargs: Dict,
+                  tensor_input_allocs: List = None) -> List:
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py:70]: update return type (Tuple), also in function description

File: model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py [hunk 6]
```diff
@@ -125,14 +133,15 @@ def _run_operation(n: BaseNode,
...
         input_tensors = [to_torch_tensor(t, numpy_type=t.dtype) if isinstance(t, np.ndarray) else t
                          for t in input_tensors]
-        _tensor_input_indices = None
+        _tensor_input_allocs = None
     else:
-        _tensor_input_indices = [i for i in n.tensor_input_indices if i not in n.weights]
+        _tensor_input_allocs = [i for i in n.tensor_input_allocs if i not in n.weights]
 
     if isinstance(n, FunctionalNode) and n.inputs_as_list:
         out_tensors_of_n_float = op_func(input_tensors, *op_call_args, **functional_kwargs)
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py:141]: why if input_as_list then we don't need to merge inputs?

File: tests/pytorch_tests/model_tests/feature_models/const_representation_test.py [hunk 7]
```diff
@@ -154,3 +154,56 @@ def __init__(self, unit_test, func, const, indices):
...
+        # reshaping batch_norm input to 3 axes to avoid bn-folding.
+        x = x.reshape((-1, 16, int(np.prod(self.input_shape))))
+
+        # input const in kwargs (not the first kwargs!)
+        x = nn.functional.batch_norm(x,
+                                     self.bn.running_mean, self.bn.running_var,
+                                     momentum=0.2, eps=1e-6, bias=self.bn.bias)
+
+        # input all tensors and consts in kwargs
+        x = torch.sub(input=self.sub_const, other=x)
```
Comment (ofirgo) [inline tests/pytorch_tests/model_tests/feature_models/const_representation_test.py:183]: maybe also add to the model a functional op without key arguments? (just x+y or something like that)

File: tests/pytorch_tests/model_tests/feature_models/const_representation_test.py [hunk 8]
```diff
@@ -154,3 +154,56 @@ def __init__(self, unit_test, func, const, indices):
...
+        return generate_pytorch_tpc(name="linear_collapsing_test", tp_model=tp)
+
+    def compare(self, quantized_model, float_model, input_x=None, quantization_info=None):
+        in_torch_tensor = to_torch_tensor(input_x[0])
+        set_model(float_model)
+        y = float_model(in_torch_tensor)
+        y_hat = quantized_model(in_torch_tensor)
+        self.unit_test.assertTrue(y.shape == y_hat.shape, msg=f'out shape is not as expected!')
+        cs = cosine_similarity(torch_tensor_to_numpy(y), torch_tensor_to_numpy(y_hat))
+        self.unit_test.assertTrue(np.isclose(cs, 1), msg=f'fail cosine similarity check:{cs}')
```
Comment (ofirgo) [inline tests/pytorch_tests/model_tests/feature_models/const_representation_test.py:209]: run this like 1000 times and check if this needs some larger tolerance

**Review by ofirgo — CHANGES_REQUESTED** [review]

File: tests/pytorch_tests/model_tests/feature_models/const_representation_test.py (hunk 7 above)
Comment (elad-c) [reply tests/pytorch_tests/model_tests/feature_models/const_representation_test.py:183]: We have other specific tests for that.

File: model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py (hunk 6 above)
Comment (elad-c) [reply model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py:141]: It is merged in the code above. I'll fix it in the next refactor PR.

File: model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py (hunk 2 above)
Comment (elad-c) [reply model_compression_toolkit/core/pytorch/back2framework/pytorch_model_builder.py:84]: I'll fix it in the next refactor PR, when handling the `inputs_as_list` comment.

**Review by ofirgo — APPROVED** [review]
